core/tee/tadb.c

```diff
@@ -0,0 +1,702 @@
+{
+	void *p;
+
+	if (idx < db->nbits)
+		return TEE_SUCCESS;
+
+	p = realloc(db->files, bitstr_size(idx + 1));
+	if (!p)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	db->files = p;
+
+	bit_nclear(db->files, db->nbits, idx);
+	db->nbits = idx + 1;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result set_file(struct tee_tadb_dir *db, int idx)
+{
+	TEE_Result res = maybe_grow_files(db, idx);
+
+	if (!res)
+		bit_set(db->files, idx);
+
+	return res;
+}
+
+static void clear_file(struct tee_tadb_dir *db, int idx)
+{
+	if (idx < db->nbits)
+		bit_clear(db->files, idx);
+}
+
+static bool test_file(struct tee_tadb_dir *db, int idx)
+{
+	if (idx < db->nbits)
+		return bit_test(db->files, idx);
+
+	return false;
+}
+
+static TEE_Result read_ent(struct tee_tadb_dir *db, size_t idx,
+			   struct tadb_entry *entry)
+{
+	size_t l = sizeof(*entry);
+	TEE_Result res = db->ops->read(db->fh, idx * l, entry, &l);
+
+	if (!res && l != sizeof(*entry))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	return res;
+}
+
+static TEE_Result write_ent(struct tee_tadb_dir *db, size_t idx,
+			    const struct tadb_entry *entry)
+{
+	const size_t l = sizeof(*entry);
+
+	return db->ops->write(db->fh, idx * l, entry, l);
+}
+
+static TEE_Result tadb_open(struct tee_tadb_dir **db_ret)
+{
+	TEE_Result res;
+	struct tee_tadb_dir *db = calloc(1, sizeof(*db));
+	struct tee_pobj po = {
+		.obj_id = (void *)tadb_obj_id,
+		.obj_id_len = sizeof(tadb_obj_id)
+	};
+
+	if (!db)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	db->ops = tee_svc_storage_file_ops(TEE_STORAGE_PRIVATE);
+
+	res = db->ops->open(&po, NULL, &db->fh);
+	if (res == TEE_ERROR_ITEM_NOT_FOUND)
+		res = db->ops->create(&po, false, NULL, 0, NULL, 0, NULL, 0,
+				      &db->fh);
+
+	if (res) {
+		free(db);
+	} else {
+		db->refc++;
+		*db_ret = db;
+	}
+
+	return res;
+}
+
+static struct tee_tadb_dir *tadb_get(struct tee_tadb_dir *db)
+{
+	assert(db->refc);
+	db->refc++;
+	assert(db->refc);
+
+	return db;
+}
+
+static TEE_Result tee_tadb_open(struct tee_tadb_dir **db)
+{
+	if (!tadb_db) {
+		TEE_Result res = tadb_open(&tadb_db);
+
+		if (res)
+			return res;
+	} else
+		tadb_get(tadb_db);
+
+	*db = tadb_db;
+	return TEE_SUCCESS;
+}
+
+static void tadb_put(struct tee_tadb_dir *db)
+{
+	assert(db == tadb_db);
+	if (!db)
+		return;
+
+	assert(db->refc);
+	db->refc--;
+	if (db->refc)
+		return;
+
+	db->ops->close(&db->fh);
+	free(db);
+	tadb_db = NULL;
+}
+
+static void tee_tadb_close(struct tee_tadb_dir *db)
+{
+	tadb_put(db);
+}
+
+static TEE_Result populate_files(struct tee_tadb_dir *db)
+{
+	TEE_Result res;
+	size_t idx;
+	const TEE_UUID null_uuid = { 0 };
+
+	if (db->files)
+		return TEE_SUCCESS;
+
+
+	for (idx = 0;; idx++) {
+		struct tadb_entry entry;
+
+		res = read_ent(db, idx, &entry);
+		if (res) {
+			if (res == TEE_ERROR_ITEM_NOT_FOUND)
+				return TEE_SUCCESS;
+			break;
+		}
+
+		if (!memcmp(&entry.prop.uuid, &null_uuid, sizeof(null_uuid)))
+			continue;
+
+		if (test_file(db, entry.file_number)) {
```

Comment:
... and there you use `db->files` to check if `file_number` is stored. But at L265 you ensured that `db->files` is `NULL`.

Reply:
As this is called in a loop, `db->files` could have been allocated at this point. `test_file()` and `set_file()` deals with that.